src/manifest.rs

```diff
@@ -0,0 +1,460 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fs::{self, File, OpenOptions},
+    io::{BufReader, Read, Seek, SeekFrom, Write},
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+
+use bytes::{Buf, BufMut, BytesMut};
+use crc::crc32;
+use prost::Message;
+use proto::meta::{
+    manifest_change::Operation as ManifestChangeOp, ManifestChange, ManifestChangeSet,
+};
+
+use crate::{util, AgateOptions, Error, Result};
+
+pub const MANIFEST_FILENAME: &str = "MANIFEST";
+const MANIFEST_REWRITE_FILENAME: &str = "MANIFEST_REWRITE";
+const MANIFEST_DELETION_REWRITE_THRESHOLD: usize = 10000;
+const MANIFEST_DELETIONS_RATIO: usize = 10;
+
+const MAGIC_TEXT: &[u8] = b"Agat";
+const MAGIC_VERSION: u32 = 8;
+
+/// LevelManifest contains information about LSM tree levels
+/// in the MANIFEST file.
+#[derive(Default, Clone, Debug)]
+pub struct LevelManifest {
+    pub tables: HashSet<u64>,
+}
+
+/// TableManifest contains information about a specific table
+/// in the LSM tree.
+#[derive(Clone, Debug)]
+pub struct TableManifest {
+    pub level: u8,
+    pub key_id: u64,
+    // TODO: compression
+}
+
+/// Manifest represents the contents of the MANIFEST file.
+#[derive(Clone, Debug)]
+pub struct Manifest {
```

Comment:
where it is used

Reply:
It will be used in `db` and `levels` modules, which are not implemented yet.